src/mscorlib/shared/System/Memory.cs

```diff
@@ -0,0 +1,313 @@
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory<T> Slice(int start)
+        {
+            if ((uint)start > (uint)_length)
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            // There is no need to 'and' _index by the bit mask here 
+            // since the constructor will set the highest order bit again anyway
+            if (_index < 0)
+                return new Memory<T>(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory), _index + start, _length - start);
+            return new Memory<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index + start, _length - start);
+        }
+
+        /// <summary>
+        /// Forms a slice out of the given memory, beginning at 'start', of given length
+        /// </summary>
+        /// <param name="start">The index at which to begin this slice.</param>
+        /// <param name="length">The desired length for the slice (exclusive).</param>
+        /// <exception cref="System.ArgumentOutOfRangeException">
+        /// Thrown when the specified <paramref name="start"/> or end index is not in range (&lt;0 or &gt;=Length).
+        /// </exception>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public Memory<T> Slice(int start, int length)
+        {
+            if ((uint)start > (uint)_length || (uint)length > (uint)(_length - start))
+                ThrowHelper.ThrowArgumentOutOfRangeException();
+
+            // There is no need to 'and' _index by the bit mask here 
+            // since the constructor will set the highest order bit again anyway
+            if (_index < 0)
+                return new Memory<T>(Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory), _index + start, length);
+            return new Memory<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index + start, length);
+        }
+
+        /// <summary>
+        /// Returns a span from the memory.
+        /// </summary>
+        public Span<T> Span
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                if (_index < 0)
+                    return Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).AsSpan(_index & bitMask, _length);
+                return new Span<T>(Unsafe.As<T[]>(_arrayOrOwnedMemory), _index, _length);
+            }
+        }
+
+        public unsafe MemoryHandle Retain(bool pin = false)
+        {
+            MemoryHandle memoryHandle;
+            if (pin)
+            {
+                if (_index < 0)
+                {
+                    memoryHandle = Unsafe.As<OwnedMemory<T>>(_arrayOrOwnedMemory).Pin();
+                }
+                else
+                {
+                    var handle = GCHandle.Alloc(Unsafe.As<T[]>(_arrayOrOwnedMemory), GCHandleType.Pinned);
+                    void* pointer = Unsafe.Add<T>((void*)handle.AddrOfPinnedObject(), _index);
```

Comment:
I'd previously seen other methods of getting the address be a bit faster:
https://github.com/dotnet/corefx/pull/22170#discussion_r127308288
Just wondering if you looked at other approaches here (and maybe it doesn't matter).

Reply:
I had not looked at the other approaches. This is interesting. I will try what Jan suggested.

Reply:
The two approaches below do not give the same memory address:

```C#
var array = new byte[1000];
var handle = GCHandle.Alloc(array, GCHandleType.Pinned);
void* pointer1 = (void*)handle.AddrOfPinnedObject();
void* pointer2 = Unsafe.AsPointer(ref handle);
Console.WriteLine(new IntPtr(pointer1) + ":" + new IntPtr(pointer2)); // Output: 2201489332176:670203108448
```

Reply:
> The two approaches below do not give the same memory address

Why would they?  The address of the handle is different from the address wrapped by the handle.

Reply:
Hmm. That makes sense. But then how can we get the address without using AddrOfPinnedObject()?

Reply:
AddrOfPinnedObject might be perfectly fine.  But other approaches you could try would be:
```C#
void* pointer;
fixed (void* tmp = array) pointer = tmp;
```
or
```C#
void* pointer = array.Length > 0 ?
    Unsafe.AsPointer(ref array[0]) :
    null;
```
or:
```C#
void* pointer = Marshal.UnsafeAddrOfPinnedArrayElement(array, 0);
```
etc.

Reply:
**Question:** Do approaches 2 and 3 work because the array has been pinned by our GCHandle.Alloc call and will not get moved by the GC? If we don't have the `GCHandle.Alloc(array, GCHandleType.Pinned);` call and get the pointer without using `fixed`, the array could still move, correct?

Reply:
Correct